lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+        // The current FieldNum has to be an array to be a padding
+        if (dyn_cast<DIType>(T)->getTag() ==
+            dwarf::Tag::DW_TAG_structure_type) {
+          assert(isa<DICompositeType>(T));
+        }
+        DEBUG_WITH_TYPE(DEBUG_TYPE,
+                        dbgs() << "Field " << FieldNum + 1
+                               << " might be a padding because its size is "
+                               << FieldSizes[FieldNum]
+                               << " Bytes but the actual field is " << Size
+                               << " Bytes\n");
+        // Mark this FieldNum as a padding
+        FieldDI[FieldNum] = nullptr;
+        // Move on the next FieldNum to check if we can find a match
+        FieldNum++;
+        Size = dyn_cast<DIType>(T)->getSizeInBits() / 8;
+      }
+      // There has to be a FieldNum that can match the size of current field
+      // in debug info
+      assert(FieldNum < NumElements);
+      DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER, dbgs()
+                                              << *dyn_cast<DIType>(T) << "\n");
+      FieldDI[FieldNum] =
+          new FieldDebugInfo(i, dyn_cast<DIDerivedType>(node)->getName(),
+                             dyn_cast<DIType>(T)->getName());
+      FieldNum++;
+    }
+  } else {
+    unsigned Address;
+    unsigned FieldNum = 0;
+    for (unsigned i = 0; i < NumElements; i++) {
+      auto *Ty = StructureType->getElementType(i);
+      if (!Ty->isArrayTy()) {
+        FieldDI[i] = new FieldDebugInfo(FieldNum++);
+        Address += FieldSizes[i];
+      } else {
+        // ArrayTy is suspicious to be a padding
+        if ((i < NumElements - 1 && FieldSizes[i] > FieldSizes[i + 1]) ||
+            (i == NumElements - 1 && FieldSizes[i] > FieldSizes[i - 1]))
+          // Padding can't be larger than the field before or after
+          FieldDI[i] = new FieldDebugInfo(FieldNum++);
+        else if ((i < NumElements - 1 && Address % FieldSizes[i + 1] != 0 &&
+                  (Address + FieldSizes[i]) % FieldSizes[i + 1] == 0) ||
+                 (i == NumElements - 1 && Address % FieldSizes[i - 1] != 0 &&
+                  (Address + FieldSizes[i]) % FieldSizes[i - 1] == 0)) {
+          // If remove this field, the next/previous field is not aligned and
+          // add this field, it is aligned, it's likely to be a padding
+          unsigned MaxWCP = 0;
+          dbgs() << "Considering F" << i + 1 << "\n";
+          for (unsigned j = 0; j < NumElements; j++) {
+            auto *Pair = CPBuilder->getCloseProximityPair(i, j);
+            dbgs() << "(" << Pair->first << "," << Pair->second << ")\n";
+            if (Pair->first > MaxWCP)
+              MaxWCP = Pair->first;
+          }
+          if (MaxWCP != 0)
+            FieldDI[i] = new FieldDebugInfo(FieldNum++);
+          else
+            // This is highly likely to be a padding because there's no WCP with
+            // other fields
+            FieldDI[i] = nullptr;
```

Comment:
You should actually be checking if there is any access to the field to determine if it could possibly be a padding. If that is hard to thread, then fine, use MaxWCP. In any case, that should be the first check.